### PR TITLE
Fix #5: set up scripts architecture

### LINKED ISSUE #5: Set up "plugin" architecture for multiple bot scripts
When implementing #4, make the script be part of some list of plugins that process the chat messages. That way, multiple scripts can be developed without stepping on each others' toes.

### PR DESCRIPTION
**Pull request by avik-das**
In doing so, I made use of Python 3.5's `async`/`await` functionality,
which allowed me to refactor some of the existing code to be
self-consistent.  To add new scripts, the steps are:

- Add a new file in the `scripts/` directory. This file should have an
  `async` function named `process`. The function should return `True` if
  the message was processed and `False` otherwise.
- Import this module in `__main__.py` and add it to the
  `ENABLED_SCRIPTS` array.

As part of this change, I added the first script to send random Ron
Swanson quotes.

Fixes #5.

@tnterdan: Can you please take a look? I wanted to get this architecture in place so you could isolate your giphy script and include it as a standalone script.

### CODE REVIEW

File: __main__.py [hunk 1]
```diff
@@ -86,38 +93,29 @@ def conversation_name(conv):
...
-    return receive_message
+    try:
+        for script in ENABLED_SCRIPTS:
+            if await script.process(client, conversation, message_event):
+                break
+    except Exception as e:
+        logger.exception('Failed to send process message: %s', e)
 
 
 def on_event_handler(loop, client, conversation):
```
Comment (tnterdan) [inline __main__.py]: No need to send the loop anymore.

File: __main__.py [hunk 2]
```diff
@@ -86,38 +93,29 @@ def conversation_name(conv):
...
+async def receive_message(message_event, client, conversation):
+    """
+    For a chat message event, go through the enabled scripts until one of
+    them processes a message.
+    """
 
-    return receive_message
+    try:
+        for script in ENABLED_SCRIPTS:
+            if await script.process(client, conversation, message_event):
```
Comment (tnterdan) [inline __main__.py]: I think it could be a little clearer to name it `process_message` and then order the arguments in the same way as `receive_message`.

**Review by tnterdan — CHANGES_REQUESTED** [review]

File: scripts/swanson.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+async def process(client, conversation, message_event):
+    if MESSAGE_REGEX.match(message_event.text):
+        loop = asyncio.get_event_loop()
+        response = await loop.run_in_executor(None, requests.get, API_URL)
+        quote = response.json()[0]
+
+        segments = hangups.ChatMessageSegment.from_str(
+            '"{}"\n- Ron Swanson'.format(quote))
+        conversation.send_message(segments)
```
Comment (tnterdan) [inline scripts/swanson.py]: Why is this here twice?

Comment (avik-das) [reply scripts/swanson.py]: No idea. Probably related to changing around the async structure heavily as I was figuring out asyncio. I'll remove the non-awaited version.

File: __main__.py (hunk 2 above)
Comment (avik-das) [inline __main__.py]: True. I was envisioning an architecture that was could handle other types of events, but (a) we should probably limit scripts to messages and maybe one or two other types of events, and (b) it'll be good for the dispatch to be done in one central location instead of each script. So I like the idea of `process_message`.

As for the arguments order, I'm going to change `receive_message` to have the same order as `process_message`. One great piece of advice I got was to put more general, contextual parameters earlier and more specific ones later. (This is great in languages that support partial application, since you can bind earlier parameters and end up with functions that operate on very specific pieces of data.)

File: __main__.py (hunk 1 above)
Comment (avik-das) [reply __main__.py]: Good catch, and a nice improvement to the architecture as well!

File: __main__.py (hunk 2 above)
Comment (avik-das) [reply __main__.py]: True. I was envisioning an architecture that was could handle other types of events, but (a) we should probably limit scripts to messages and maybe one or two other types of events, and (b) it'll be good for the dispatch to be done in one central location instead of each script. So I like the idea of `process_message`.

As for the arguments order, I'm going to change `receive_message` to have the same order as `process_message`. One great piece of advice I got was to put more general, contextual parameters earlier and more specific ones later. (This is great in languages that support partial application, since you can bind earlier parameters and end up with functions that operate on very specific pieces of data.)

**Comment by avik-das** [comment]
@tnterdan: Addressed your feedback. Can you take another look please?

**Review by tnterdan — APPROVED** [review]
